src/Core/ExecutionPathTracer/ExecutionPathTracer.cs

```diff
@@ -0,0 +1,307 @@
+        {
+            var qubits = this.qubitRegisters.Keys
+                .OrderBy(k => k)
+                .Select(k =>
+                {
+                    var qubitDecl = new QubitDeclaration(k);
+                    if (this.classicalRegisters.ContainsKey(k))
+                    {
+                        qubitDecl.NumChildren = this.classicalRegisters[k].Count;
+                    }
+
+                    return qubitDecl;
+                })
+                .ToArray();
+
+            return new ExecutionPath(qubits, this.operations.ToArray());
+        }
+
+        /// <summary>
+        /// Provides the event listener to listen to <c>SimulatorBase</c>'s <c>OnOperationStart</c> event.
+        /// </summary>
+        public void OnOperationStartHandler(ICallable operation, IApplyData arguments)
+        {
+            // If the operation type is one of the nestedTypes, go one depth deeper and parse
+            // those operations instead
+            if (this.nestedTypes.Contains(operation.GetType())) return;
+
+            this.currDepth++;
+
+            // Parse operations at specified depth
+            if (this.currDepth == this.renderDepth)
+            {
+                var parsedOp = this.ParseOperation(operation, arguments);
+                if (parsedOp != null)
+                {
+                    this.operations.Add(parsedOp);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Provides the event listener to listen to <c>SimulatorBase</c>'s <c>OnOperationEnd</c> event.
+        /// </summary>
+        public void OnOperationEndHandler(ICallable operation, IApplyData result)
+        {
+            if (this.nestedTypes.Contains(operation.GetType())) return;
+            this.currDepth--;
+        }
+
+        private static bool IsPartialApplication(ICallable operation)
+        {
+            var t = operation.GetType();
+            if (t == null) return false;
+
+            return t.IsGenericType && t.GetGenericTypeDefinition() == typeof(OperationPartial<,,>);
+        }
+
+        /// <summary>
+        /// Retrieves the <c>QubitRegister</c> associated with the given <c>Qubit</c> or create a new
+        /// one if it doesn't exist.
+        /// </summary>
+        private QubitRegister GetQubitRegister(Qubit qubit)
+        {
+            if (!this.qubitRegisters.ContainsKey(qubit.Id))
+            {
+                this.qubitRegisters[qubit.Id] = new QubitRegister(qubit.Id);
+            }
+
+            return this.qubitRegisters[qubit.Id];
+        }
+
+        private List<QubitRegister> GetQubitRegisters(IEnumerable<Qubit> qubits)
+        {
+            return qubits.Select(this.GetQubitRegister).ToList();
+        }
+
+        /// <summary>
+        /// Creates a new <c>ClassicalRegister</c> and associate it with the given <c>Qubit</c>.
+        /// </summary>
+        private ClassicalRegister CreateClassicalRegister(Qubit measureQubit)
```

Comment:
This feels similar enough to the previous that you really want an extension method on `IDictionary<TKey, TValue>` that gets or creates a key.

Reply:
Oh, interesting, that would make sense. Thanks!